core/storage/blog_statistics/gae_models_test.py

```diff
@@ -0,0 +1,933 @@
+    NONEXISTENT_USER_ID: Final = 'user_id_x'
+
+    BLOG_POST_ONE_ID: Final = 'blog_post_one'
+    BLOG_POST_TWO_ID: Final = 'blog_post_two'
+
+    def setUp(self) -> None:
+        """Set up blog post view even log entry models in datastore for use in
+        testing.
+        """
+        super().setUp()
+
+        self.event_model_one = (
+            blog_stats_models.BlogPostViewedEventLogEntryModel(
+                id='event_model_1',
+                blog_post_id=self.BLOG_POST_ONE_ID
+            )
+        )
+        self.event_model_one.update_timestamps()
+        self.event_model_one.put()
+
+        self.event_model_two = (
+            blog_stats_models.BlogPostViewedEventLogEntryModel(
+                id='event_model_2',
+                blog_post_id=self.BLOG_POST_ONE_ID
+            )
+        )
+        self.event_model_two.update_timestamps()
+        self.event_model_two.put()
+
+        self.event_model_three = (
+            blog_stats_models.BlogPostViewedEventLogEntryModel(
+                id='event_model_3',
+                blog_post_id=self.BLOG_POST_ONE_ID
+            )
+        )
+        self.event_model_three.update_timestamps()
+        self.event_model_three.put()
+
+        self.event_model_four = (
+            blog_stats_models.BlogPostViewedEventLogEntryModel(
+                id='event_model_4',
+                blog_post_id=self.BLOG_POST_TWO_ID
+            )
+        )
+        self.event_model_four.update_timestamps()
+        self.event_model_four.put()
+
+    def test_get_deletion_policy(self) -> None:
+        self.assertEqual(
+            blog_stats_models.BlogPostViewedEventLogEntryModel
+            .get_deletion_policy(),
+            base_models.DELETION_POLICY.NOT_APPLICABLE
+        )
+
+    def test_create_raises_exception_by_mocking_collision(self) -> None:
+        stats_model_class = blog_stats_models.BlogPostViewedEventLogEntryModel
+        with self.assertRaisesRegex(
+            Exception, 'The id generator for the model is producing too many'
+            ' collisions.'):
+            # Swap dependent method get_by_id to simulate collision every time.
+            with self.swap(
+                blog_stats_models.BlogPostViewedEventLogEntryModel,
+                'get_by_id', types.MethodType(
+                    lambda x, y: True,
+                    blog_stats_models.BlogPostViewedEventLogEntryModel)):
+                stats_model_class.create('blog_post_id')
+
+    def test_get_new_event_entity_id(self) -> None:
+        time_instance1 = utils.get_current_time_in_millisecs()
+        time.sleep(1)
+        new_id = (
+            blog_stats_models.BlogPostViewedEventLogEntryModel
+                .get_new_event_entity_id('BLOG')
+        )
+        time_id_created = int(new_id.split(':')[0])
+        self.assertGreater(time_id_created, time_instance1)
+        time.sleep(1)
+        self.assertLess(time_id_created, utils.get_current_time_in_millisecs())
```

Comment:
Instead of asserting less, I would recomment mocking this function to return a specific floating timestamp. Similarly, you can mock your hash function to return a specific hash. Then assert that final generated id is equal to your test value.

Reply:
Done

Reply:
I think you should assert the whole if at once? Will be easier for authors to review the outcome of the test

Reply:
Done